File: 62-unique-paths/62-unique-paths.java
class Solution {
    public int uniquePaths(int m, int n) {
         return fun(m,n,0,0, new HashMap<String,Integer>());
    }
    
    public int fun(int m,int n, int r, int c, HashMap<String,Integer> memo){
    
    if(r>=m)return 0;
    if(c>=n)return 0;
    
    if(r==m-1&&c==n-1)return 1;
        String currkey=Integer.toString(r)+'_'+Integer.toString(c);
        if(memo.containsKey(currkey)){
            return memo.get(currkey);
        }
    int right=fun( m, n, r,  c+1,memo);
    int down=fun( m, n, r+1,  c,memo);
    memo.put(currkey,right+down);
    return right + down;
    }
    
}

File: 62-unique-paths/62-unique-paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>>dp(m+1,vector<int>(n+1,-1));
        return totalWays(dp,m,n,0,0);
    }
    
    int totalWays(vector<vector<int>>&dp,int m,int n,int row,int col){
        if(row<0||row>=m||col<0||col>=n)
            return 0;
        if(row==m-1&&col==n-1)
            return 1;
        
        if(dp[row][col]!=-1)
            return dp[row][col];
        
        int down=totalWays(dp,m,n,row+1,col);
        int right=totalWays(dp,m,n,row,col+1);
        return dp[row][col]=down+right;
    }
};